icicle-compiler/src/Icicle/Command.hs

```diff
@@ -0,0 +1,370 @@
+        InputSparsePsv ->
+          FormatPsv $ psvDefaultConstants
+            (PsvInputConfig Chords PsvInputSparse)
+            (PsvOutputConfig Chords PsvOutputSparse defaultOutputMissing)
+        InputDensePsv ->
+          FormatPsv $ psvDefaultConstants
+            (PsvInputConfig Chords PsvInputSparse)
+            (PsvOutputConfig Chords PsvOutputSparse defaultOutputMissing)
+        InputZebra ->
+          FormatZebra Chords
+            (PsvOutputConfig Chords PsvOutputSparse defaultOutputMissing)
+
+  let cfg = HasInput format (InputOpts AllowDupTime Map.empty) (inputPath input)
+  code  <- liftIO $ Text.readFile source
+  fleet <- firstEitherT IcicleSeaError (seaCreate CacheSea cfg chord code)
+  return (code, fleet)
+
+compileFleet ::
+     Dictionary
+  -> IOFormat
+  -> FilePath
+  -> Maybe FilePath
+  -> EitherT IcicleError IO (Text, SeaFleet s)
+compileFleet dictionary format input chords = do
+  let cfg = HasInput format (InputOpts AllowDupTime (tombstonesOfDictionary dictionary)) input
+
+  avalanche <- hoistEither $ compileAvalanche dictionary
+  let avalancheL = Map.toList avalanche
+
+  code  <- firstEitherT IcicleSeaError (hoistEither (codeOfPrograms cfg avalancheL))
+  fleet <- firstEitherT IcicleSeaError (seaCompile CacheSea cfg avalanche chords)
+
+  return (code, fleet)
+
+loadDictionary ::
+     FilePath
+  -> InputFormat
+  -> OutputFormat
+  -> Scope a
+  -> EitherT IcicleError IO (Dictionary, IOFormat)
+loadDictionary path iformat oformat0 scope =
+  let
+    mode =
+      modeOfScope scope
+
+    oformat =
+      psvOfOutputFormat oformat0
+
+    oconfig =
+      PsvOutputConfig mode oformat defaultOutputMissing
+  in
+    case iformat of
+      InputSparsePsv -> do
+        d <- firstEitherT IcicleDictionaryImportError $
+          Toml.loadDictionary Source.optionSmallData Toml.ImplicitPrelude path
+
+        let f = FormatPsv $ psvDefaultConstants
+                  (PsvInputConfig mode PsvInputSparse)
+                  oconfig
+
+        return (d, f)
+
+      InputDensePsv -> do
+        (d, dense) <- firstEitherT IcicleDictionaryImportError $
+          Toml.loadDenseDictionary Source.optionSmallData Toml.ImplicitPrelude path Nothing
+
+        let f = FormatPsv $ psvDefaultConstants
+                  (PsvInputConfig mode (PsvInputDense dense (denseSelectedFeed dense)))
+                  oconfig
+
+        return (d, f)
+
+      InputZebra -> do
+        d <- firstEitherT IcicleDictionaryImportError $
+          Toml.loadDictionary Source.optionSmallData Toml.ImplicitPrelude path
+
+        let f = FormatZebra mode oconfig
+
+        return (d, f)
+
+compileDictionary ::
+     FilePath
+  -> InputFormat
+  -> OutputFormat
+  -> Scope a
+  -> EitherT IcicleError IO Text
+compileDictionary dictionaryPath iformat oformat scope = do
+  -- FIXME We really need to include InputFormat/OutputFormat/Scope in the compiled
+  -- FIXME code so that we don't accidentally run with the wrong options.
```

Comment:
yeah, or maybe better, not actually output C code, but a tarball with the C code and a `.json` file that describes the options a file was compiled with. Then we can check that we're not going to explode when we try to run the snapshot.